Add custom visitor cookie timeout setting

tomper00 opened the pull request:
Add possibility to set custom a custom cookie timeout in the Matomo config variable as described in https://github.com/matomo-org/tag-manager/issues/562

### Description:

This will add a new setting in the Matomo Config variable where you can manually set the timeout in seconds for the pk_id cookie. 
 

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

tomper00 committed:
Add custom visitor cookie timeout setting

Add possibility to set custom a custom cookie timeout in the Matomo config variable as described in https://github.com/matomo-org/tag-manager/issues/562

snake14 reviewed (changes requested):
Hi @tomper00 . Thank you for taking the time to create this PR. I think it looks pretty good, but there are a few changes I'd like to request.
1. Could you please update the tests that are broken by your changes? If you need help with the screenshots or any other parts, just let me know.
2. Could you please [create a migration to add your new fields to existing tags](https://developer.matomo.org/guides/tagmanager/parameters#migrating-tags-to-include-new-parameters)?
3. Could you please [add some validation](https://developer.matomo.org/guides/tagmanager/parameters#validating-a-value) to the `customCookieTimeOut` field. The typing handles some validation, but I think a minimum value would be a good idea to at least prevent 0 and negative values.

snake14 commented:
@mattab This customer provided PR is looking really good and I'd like to merge it after a few additions/adjustments. Do [the new fields](https://github.com/matomo-org/tag-manager/pull/563/files#diff-d95e907677433b4141e187dd98997bef7e9479abf9055e31bc1012e0e8901f42) look alright to you and do the [labels/wording](https://github.com/matomo-org/tag-manager/pull/563/files#diff-1d56632a2d162f99c901e53b79a6224e984c14501bca3188a85dbb4b4c5e6da1) look alright?

tomper00 committed:
PR Improvements

Updated failed tests
Added validation to customCookieTimeOut
Added tag migration script (hopefully got this right=

tomper00 committed:
Add num range validator

Add num range validator

tomper00 committed:
Change order in tests

snake14 commented on Updates/4.12.4-b1.php:
```diff
@@ -0,0 +1,75 @@
+    public function doUpdate(Updater $updater)
+    {
+        // Migrate the Matomo type tags to all include the newly configured field.
+        $migrator = new NewTagParameterMigrator(MatomoTag::ID, 'someNewParameterName',0);
```
Thank you for making this migration @tomper00 . It looks good. Please replace `someNewParameterName` with `customCookieTimeOutEnable`. I also think that the 0 default value should probably be `false`.

snake14 commented on tests/System/expected/test___TagManager.exportContainerVersion_site_default_container.xml:
```diff
@@ -88,6 +88,8 @@
 				<requireConsent>0</requireConsent>
 				<requireCookieConsent>0</requireCookieConsent>
+				<customCookieTimeOutEnable>0</customCookieTimeOutEnable>
+				<customCookieTimeOut>0</customCookieTimeOut>
```
Looks like the failing test is expecting `33955200` instead of `0`. Please update.

snake14 commented on tests/System/expected/test_webContext__TagManager.getAvailableVariableTypesInContext.xml:
```diff
@@ -1005,6 +1005,40 @@
+						<title>The pk_id expire time in seconds</title>
+						<value>33955200</value>
+						<defaultValue>33955200</defaultValue>
+						<type>int</type>
```
Looks like the failing test is expecting `integer` instead of `int`. Please update.

snake14 reviewed (changes requested):
@tomper00 Thank you again for all your effort on this PR. I made a few comments about minor changes. After those are made, I think that this PR will be ready as long as our product team is alright with the label text.

tomper00 committed:
Adjust tests and update script

Based on great feedback from snake14

tomper00 committed:
Change setting to use full days

Make it easier to configure by using full days instead of seconds

tomper00 committed:
Fix test for new default data

393 days instead of 33955200 seconds

tomper00 committed:
Update test_webContext__TagManager.getAvailableVariableTypesInContext.xml

Fix tests with new defaults

tomper00 commented:
@snake14 Great feedback, thanks for the help!
I added the new tests and also changed the variable to use days instead of seconds, since I think that is a much better from a usability perspective. 
Looks like the tests are ok now.

snake14 commented on Updates/4.12.4-b1.php:
```diff
@@ -0,0 +1,75 @@
+    {
+        // Migrate the Matomo type tags to all include the newly configured field.
+        $migrator = new NewTagParameterMigrator(MatomoTag::ID, 'customCookieTimeOutEnable',0);
+        $migrator->addField('customCookieTimeOut',33955200); // This is optional and only needed if you added more than one parameter.
```
I believe that this should also be a value of `393` to be consistent with your changes.

snake14 reviewed (changes requested):
@tomper00 Things are looking good. I only had one more requested change after your change from seconds to days. My only other thought is that some customers might want more granularity than a day, but what you've got should work.

tomper00 committed:
Fix default value in update script

Fix default value in update script to match full days!

tomper00 commented:
Good point with the granularity, but it's probably good enough for now. I updated the update script default value now as well, thanks again for the feedback!

snake14 commented on Updates/4.12.4-b1.php:
```diff
@@ -0,0 +1,75 @@
+        $migrator = new NewTagParameterMigrator(MatomoTag::ID, 'customCookieTimeOutEnable',0);
+        $migrator->addField('customCookieTimeOut',393); // This is optional and only needed if you added more than one parameter.
+        $migrator->migrate(); // This kicks off the processing of the tag migration.
+    }
```
@tomper00 I must apologise for not catching this sooner, but you defined the `doUpdate` method twice. Also, I slightly misled you. Since you're adding the new fields to a variable and not a tag, you need to use the `NewVariableParameterMigrator` instead of `NewTagParameterMigrator` and `MatomoConfigurationVariable` instead of `MatomoTag`. Your method should look like this:
```
    public function doUpdate(Updater $updater)
    {
        // Migrate the Matomo type tags to all include the newly configured field.
        $migrator = new NewVariableParameterMigrator(MatomoConfigurationVariable::ID, 'customCookieTimeOutEnable',0);
        $migrator->addField('customCookieTimeOut',393); // This is optional and only needed if you added more than one parameter.
        $migrator->migrate(); // This kicks off the processing of the tag migration.
    }
```

You can also get rid of everything inside the class other than the `doUpdate` method provided above. With those changes, I tested the migration and it worked as expected.

tomper00 replied:
Aha np, just so great to learn this stuff!

snake14 reviewed (changes requested):
Sorry. One more change and I think it will be ready to merge. I tested it with my suggested changes and it looks good. Please see my other comment.

tomper00 committed:
Changes to doUpdate

Changes to doUpdate after comment from @snake14

snake14 commented on Updates/4.12.4-b1.php:
```diff
@@ -0,0 +1,75 @@
+        $migrator->addField('customCookieTimeOut',393); // This is optional and only needed if you added more than one parameter.
+        $migrator->migrate(); // This kicks off the processing of the tag migration.
+    }    
+}
```
Sorry @tomper00. Looks like you've still got `doUpdate` defined twice, which results in an error. Also, don't forget to update your `use` statements up above when you change classes.

You can actually simplify the whole migration down to the following:
```
<?php
/**
 * Matomo - free/libre analytics platform
 *
 * @link https://matomo.org
 * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
 *
 */

namespace Piwik\Plugins\TagManager;

use Piwik\Plugins\TagManager\Template\Variable\MatomoConfigurationVariable;
use Piwik\Plugins\TagManager\Updates\NewVariableParameterMigrator;
use Piwik\Updater;
use Piwik\Updates as PiwikUpdates;

/**
 * Update for version 4.12.4-b1.
 */
class Updates_4_12_4_b1 extends PiwikUpdates
{
    public function doUpdate(Updater $updater)
    {
        // Migrate the Matomo type tags to all include the newly configured field.
        $migrator = new NewVariableParameterMigrator(MatomoConfigurationVariable::ID, 'customCookieTimeOutEnable',0);
        $migrator->addField('customCookieTimeOut',393);
        $migrator->migrate(); // This kicks off the processing of the tag migration.
    }
}
```

tomper00 replied:
Aha sorry for this and again tnx, pushed a new change now.

snake14 replied:
Not a problem. Thank you for all of your effort on this pull request.

tomper00 committed:
Another fix

Another fix, not reading comments good enough..

snake14 reviewed (approved):
Thank you again for all of your effort on this PR. My functional testing looked good and I think these changes look ready to merge. Does everything look alright to you @mattab ? Any label/wording changes?

AltamashShaikh commented on lang/en.json:
```diff
@@ -436,6 +436,10 @@
         "MatomoConfigurationMatomoDisableCookiesDescription": "Disables all first party cookies.",
         "MatomoConfigurationMatomoRequireConsentTitle": "Require tracking consent",
         "MatomoConfigurationMatomoRequireConsentDescription": "Track only when user gave tracking consent. In a consent screen you need to call \"window._paq=window._paq||[];window._paq.push(['rememberConsentGiven']);\" when the user gives consent.",
+        "MatomoConfigurationMatomoCustomCookieTimeOutEnableTitle": "Enable custom visitor Cookie timeout",
```
@tomper00 Can we update the text case and add a fullstop at the end to make it consistent ?
```suggestion
        "MatomoConfigurationMatomoCustomCookieTimeOutEnableTitle": "Enable Custom Visitor Cookie Timeout.",
```

AltamashShaikh commented on lang/en.json:
```diff
@@ -436,6 +436,10 @@
         "MatomoConfigurationMatomoRequireConsentTitle": "Require tracking consent",
         "MatomoConfigurationMatomoRequireConsentDescription": "Track only when user gave tracking consent. In a consent screen you need to call \"window._paq=window._paq||[];window._paq.push(['rememberConsentGiven']);\" when the user gives consent.",
+        "MatomoConfigurationMatomoCustomCookieTimeOutEnableTitle": "Enable custom visitor Cookie timeout",
+        "MatomoConfigurationMatomoCustomCookieTimeOutEnableDescription": "Allows you to set a custom time in full days for the _pk_id cookie",
```
```suggestion
        "MatomoConfigurationMatomoCustomCookieTimeOutEnableDescription": "Allows you to set a custom time in full days for the _pk_id cookie.",
```

AltamashShaikh commented on lang/en.json:
```diff
@@ -436,6 +436,10 @@
         "MatomoConfigurationMatomoRequireConsentDescription": "Track only when user gave tracking consent. In a consent screen you need to call \"window._paq=window._paq||[];window._paq.push(['rememberConsentGiven']);\" when the user gives consent.",
+        "MatomoConfigurationMatomoCustomCookieTimeOutEnableTitle": "Enable custom visitor Cookie timeout",
+        "MatomoConfigurationMatomoCustomCookieTimeOutEnableDescription": "Allows you to set a custom time in full days for the _pk_id cookie",
+        "MatomoConfigurationMatomoCustomCookieTimeOutTitle": "The pk_id expire time in full days",
```
```suggestion
        "MatomoConfigurationMatomoCustomCookieTimeOutTitle": "The pk_id expire time in full days.",
```

AltamashShaikh commented on lang/en.json:
```diff
@@ -436,6 +436,10 @@
+        "MatomoConfigurationMatomoCustomCookieTimeOutEnableTitle": "Enable custom visitor Cookie timeout",
+        "MatomoConfigurationMatomoCustomCookieTimeOutEnableDescription": "Allows you to set a custom time in full days for the _pk_id cookie",
+        "MatomoConfigurationMatomoCustomCookieTimeOutTitle": "The pk_id expire time in full days",
+        "MatomoConfigurationMatomoCustomCookieTimeOutDescription": "Timeout in full days. By default, the Matomo tracking cookie expires in 13 months (365 + 28 days) = 393 days",
```
```suggestion
        "MatomoConfigurationMatomoCustomCookieTimeOutDescription": "Timeout in full days. By default, the Matomo tracking cookie expires in 13 months (365 + 28 days) = 393 days.",
```

AltamashShaikh reviewed (commented):
@tomper00 left few textual changes, rest looks good and thanks for the PR :+1:

tomper00 committed:
Update lang/en.json

Submitting suggestions for better translations

Co-authored-by: Altamash Shaikh <altu9594@gmail.com>

tomper00 committed:
Update lang/en.json

submitting translation improvements.

Co-authored-by: Altamash Shaikh <altu9594@gmail.com>

tomper00 commented:
@AltamashShaikh thanks for helping out, your changes are submitted now.

tomper00 committed:
Merge branch 'matomo-org:4.x-dev' into 562-setVisitorCookieTimeout

tomper00 committed:
Fix tests to match new translations

Fix tests to match new translations

tomper00 committed:
Force rebuild

Force rebuild since tests failed for suspicious reasons
